src/sidebar/services/frame-sync.ts

```diff
@@ -306,9 +297,16 @@ export class FrameSyncService {
 
     guestRPC.on('close', () => {
       const frame = this._store.frames().find(f => f.id === sourceId);
-      if (frame) {
+      if (frame && !frame.persistent) {
         this._store.destroyFrame(frame);
       }
+
+      // Mark annotations as no longer being loaded in the guest, even if
+      // the frame was marked as `persistent`. In that case if a new guest
+      // connects with the same ID as the one that just went away, we'll send
+      // the already-loaded annotations to the new guest.
+      this._inFrame.clear();
```

Comment:
Note that we clear the "is this annotation in the guest frame" state for _all_ annotations here, which is what leads to the flicker mentioned in the PR description in guest frames that haven't gone away.